### PR TITLE
Adding OpenMC SS activation simulation

### LINKED ISSUE #5: Perform OpenMC activation simulation with spherical shell
Learn how to perform an activation simulation with OpenMC and extend the problem defined in #2 with the irradiation schedule defined in #4

### PR DESCRIPTION
**Pull request by anu1217**
fixes #5
- Adding OpenMC simulation
- Including .h5 files and nuclide density plot

### CODE REVIEW

File: fendl2bin.lib
Comment (gonuke) [inline fendl2bin.lib:1]: There is no need to put the data library in the repo

File: myElelib
Comment (gonuke) [inline myElelib:1]: Shouldn't need to put this element library in the repo if it's the standard one

File: statepoint.10.h5
Comment (gonuke) [inline statepoint.10.h5:1]: We should discuss which output files should be in the repo, and if so, where they should be located

File: tallies.out
Comment (gonuke) [inline tallies.out:1]: This file is probably not useful for the repo

File: SS_Activation_OpenMC.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Thu Feb  8 08:16:12 2024
+
+@author: Anupama Rajendra
+"""
+import openmc
+import openmc.deplete
+import numpy as np
+from fractions import Fraction
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: do you use this?

File: SS_Activation_OpenMC.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+neutron_tally.scores = ['flux', 'elastic', 'absorption']
+# Implementing filter for neutron tally through W shell
+cell_filter = openmc.CellFilter([Shell])
+neutron_tally.filters = [cell_filter]
+
+# Creating a tally to get the flux energy spectrum.
+# An energy filter is created to assign to the flux tally.
+e_min_1, e_max_1 = 5e2, 14.001e6
+groups = 500
+energies = np.logspace(np.log10(e_min_1), np.log10(e_max_1), groups + 1)
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: Use vitamin-J group structure here.

File: SS_Activation_OpenMC.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+energy_filter_flux = openmc.EnergyFilter(energies)
+
+e_min_2, e_max_2 = 5e2, 14.001e6
+groups = 500
+energies_2 = np.logspace(np.log10(e_min_2), np.log10(e_max_2), groups + 1)
+energy_filter_ng = openmc.EnergyFilter(energies_2)
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: This energy group structure is the same as `energies` - you don't need a different one, use the same one twice

File: SS_Activation_OpenMC.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+spectrum_tally = openmc.Tally(name="Flux spectrum")
+# Implementing energy and cell filters for flux spectrum tally
+spectrum_tally.filters = [energy_filter_flux, cell_filter]
+spectrum_tally.scores = ['flux']
```
Comment (gonuke) [inline SS_Activation_OpenMC.py:77]: Can't you just apply these filters to the `neutron_tally`?

File: SS_Activation_OpenMC.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+# Collecting and exporting tallies to .xml
+tallies = openmc.Tallies([neutron_tally, spectrum_tally])
+tallies.export_to_xml()
+
+model = openmc.model.Model(geometry=geometry,settings=settings)
+#Depletion calculation
+W.depletable = True
+W.volume = 1.33333 * np.pi * (R_2**3 - R_1**3) #volume of W wall material
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]:
```suggestion
W.volume = 4.0/3.0 * np.pi * (R_2**3 - R_1**3) #volume of W wall material
```

File: SS_Activation_OpenMC.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+chain = openmc.deplete.chain.Chain.from_xml("./chain_endfb71_sfr.xml")
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: This file name was assigned to a variable above - use that variable here to assure consistency... and below

File: Nuclide_density_OpenMC.png
Comment (gonuke) [inline Nuclide_density_OpenMC.png:1]: might need this on a log plot to see things properly - note that 5 of these are stable and won't change as a function of the shutdown time

**Review by gonuke — CHANGES_REQUESTED** [review]
This is a relatively short irradiation, with a total source of 1e20 n in a 10m radius sphere.  The uncollided flux arriving at the sphere is just under 8E10 with only 100 s of irradiation.  I would expect very little activation to occur!  You probably want multiple years of irradiation, e.g. 3e8 s for nearly 10 years.

File: SS_Activation_OpenMC.py (hunk 1 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py]: I was using this because of the 4/3 factor that appears in the spherical volume of depletable material but switched to 1.33 at some point. I now see that the Fraction module isn't necessary even when 4/3 is used -- getting rid of it

File: SS_Activation_OpenMC.py (hunk 4 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py:77]: I think I was looking for a neutron flux as a function of energy and just a total elastic/absorption score.

Comment (gonuke) [reply SS_Activation_OpenMC.py:77]: Does OpenMC give you the total over energy in addition to the spectrum when you apply an energy filter?

File: SS_Activation_OpenMC.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+#plt.plot(time, n_W181, marker='.', linestyle='solid', color='deepskyblue', label="W181")
+#plt.plot(time, n_W182, marker='.', linestyle='solid', color='peru', label="W182")
+#plt.plot(time, n_W183, marker='.', linestyle='solid', color='plum', label="W183")
+#plt.plot(time, n_W184, marker='.', linestyle='solid', color='crimson', label="W184")
+plt.plot(time, n_W185, marker='.', linestyle='solid', color='rosybrown', label="W185")
+#plt.plot(time, n_W186, marker='.', linestyle='solid', color='gray', label="W186")
+plt.plot(time, n_W187, marker='.', linestyle='solid', color='lime', label="W187")
+plt.plot(time, n_Re185, marker='.', linestyle='solid', color='gold', label="Re185")
+plt.plot(time, n_Re187, marker='.', linestyle='solid', color='tomato', label="Re187")
+plt.plot(time, n_Os187, marker='.', linestyle='solid', color='cyan', label="Os187")
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: Some kind of list data structure may be better?
```suggestion
num_dens = {}
plot_colors = {'He4' : 'red', 'W179' : 'green', 'W180' : blue } #etcetera

for nuclide, plot_color in plot_colors.items():
    time, num_dens[nuclide] =results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
    print(time, num_dens[nuclide])
    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```

File: PointSourceALARA [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+mixture W_wall
+	element  w   1.0   1.0
+end
+
+
+flux flux_3   fluxdata/flux_ss	1E+18 	0	default
+
+
+Schedule myschedule
+	100 s	  flux_3   single_pulse     0  s
```
Comment (gonuke) [inline PointSourceALARA]: Need to update this to match OpenMC

**Review by gonuke — CHANGES_REQUESTED** [review]

File: SS_Activation_OpenMC.py (hunk 4 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py:77]: Yes, it can provide a total tally in addition to the spectrum, but both have to be asked for explicitly. The total values should be the same as in PR #3 , but the spectrum may be different as we're accounting for Vitamin-J now

File: SS_Activation_OpenMC.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+"""
+import openmc
+import openmc.deplete
+import numpy as np
+import matplotlib.pyplot as plt
+import pandas as pd
+
+# Importing Vitamin-J energy group structure:
+# This excel file contains the energy bounds of the Vitamin J structure
+Vit_J = pd.read_excel('VitaminJEnergyGroupStructure.xlsx')
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: Why did you switch to XLSX?

File: SS_Activation_OpenMC.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+# Define tallies
+neutron_tally = openmc.Tally(name="Neutron tally")
+neutron_tally.scores = ['flux', 'elastic', 'absorption']
+# Implementing filter for neutron tally through W shell
+cell_filter = openmc.CellFilter([Shell])
+neutron_tally.filters = [cell_filter]
+
+# Creating a tally to get the flux energy spectrum.
+# An energy filter is created to assign to the flux tally using the Vitamin J structure.
+energy_filter_flux = openmc.EnergyFilter(ebounds)
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: I think I just learned that OpenMC has the 175 group Vitamin-J structure built in.

See the `from_group_structure` method [here](https://docs.openmc.org/en/stable/pythonapi/generated/openmc.EnergyFilter.html) and the library of structures [here](https://docs.openmc.org/en/v0.12.1/pythonapi/mgxs.html)

File: SS_Activation_OpenMC.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    nt = sp.get_tally(name="Neutron tally")
+    # Get the neutron energies from the energy filter
+    energy_filter_fl = fl.filters[0]
+    energies_fl = energy_filter_fl.bins[:, 0]
+
+    # Get the neutron flux values
+    flux = fl.get_values(value='mean').ravel()
+    
+    #Neutron flux/elastic/absorption tallies:
+    tal = nt.get_values(value='mean').ravel()
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: These lines don't have to be inside the `with` block, do they?

File: SS_Activation_OpenMC.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+Flux_Data = np.c_[energies_fl, flux]
+#Creating an excel file that stores flux data for each energy bin (used as input for ALARA)
```
Comment (gonuke) [inline SS_Activation_OpenMC.py:112]: Not sure why we want Excel?

**Review by gonuke — CHANGES_REQUESTED** [review]

File: SS_Activation_OpenMC.py (hunk 9 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py]: It was having trouble reading from the txt file and giving a tally output. For some reason it appears to work fine with the Excel file

File: SS_Activation_OpenMC.py (hunk 11 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py:102]: I think they do, or else I get an "expected an indented block" error

File: SS_Activation_OpenMC.py (hunk 10 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py]: Do you have a preference as to how the structure is implemented? (I'm guessing the built-in one)

File: SS_Activation_OpenMC.py (hunk 11 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py]: Corrected indentation

File: SS_Activation_OpenMC.py (hunk 9 above)
Comment (gonuke) [reply SS_Activation_OpenMC.py]: I'm sure we can figure out how to get the txt file version to work - certainly better than adding proprietary software files.  It may be moot for this particular case given that we will likely switch to the built in Vit-J groups.

File: SS_Activation_OpenMC.py (hunk 10 above)
Comment (gonuke) [reply SS_Activation_OpenMC.py]: I think it's best to switch to the built-in groups, yes.

File: SS_Activation_OpenMC.py (hunk 12 above)
Comment (gonuke) [reply SS_Activation_OpenMC.py:112]: Aside from my personal distaste for Excel as a data processing tool, I don't think it adds anything to make this an Excel file vs some kind of text file

File: PointSourceALARA [hunk 13]
```diff
@@ -19,21 +19,25 @@ mixture W_wall
 	element  w   1.0   1.0
 end
 
+# Normalizing neutron source rate by volume
+# This gives fluxes in #/cm^2-s when multiplied by the values in flux_ss
 
-flux flux_3   fluxdata/flux_ss	1E+18 	0	default
+flux flux_3   fluxdata/flux_ss	15836181434 	0	default
```
Comment (gonuke) [inline PointSourceALARA]: why this change?

Comment (gonuke) [reply PointSourceALARA]: OpenMC is still using 1e18

Comment (anu1217) [reply PointSourceALARA]: This is 1e18 normalized by the volume of the material, so that flux is in #/cm^2-s. There was a 6-7 order of magnitude difference in number density between OpenMC and ALARA without this

File: SS_Activation_OpenMC.py (hunk 7 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py]: Missed this comment -- will add it to the code

File: SS_Activation_OpenMC.py (hunk 10 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py]: Sounds good. Made this change and the results appear to be the same.

File: VitaminJEnergyGroupStructure.xlsx
Comment (gonuke) [inline VitaminJEnergyGroupStructure.xlsx:1]: can we delete this file now?

File: PointSourceALARA (hunk 13 above)
Comment (gonuke) [reply PointSourceALARA]: ALARA expects a flux in n/cm^2-s, so it's probably better to divide by the volume before writing the flux so that it's already in n/cm^2/source , rather than dividing manually.  That will help us see that the strengths are the same.

File: SS_Activation_OpenMC.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+energies_fl = energy_filter_fl.bins[:, 0]
+
+# Get the neutron flux values
+flux = fl.get_values(value='mean').ravel()
+
+#Neutron flux/elastic/absorption tallies:
+tal = nt.get_values(value='mean').ravel()
+print(tal)
+
+Flux_Data = np.c_[energies_fl, flux]
```
Comment (gonuke) [inline SS_Activation_OpenMC.py:111]: Here is where you may want to normalize the flux by the volume.  Strictly speaking, it's not a *flux*, per se, until you divide by the volume.

File: SS_Activation_OpenMC.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+for k in range(len(nuc_last)) :
+    Random_Color = random.choice(colors)
+    plot_colors = {nuc_last[k]: Random_Color}
+    pair_list.update(plot_colors)
+    k = k + 1
+
+for nuclide, plot_color in pair_list.items():
+    time, num_dens[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
+    print(time, num_dens[nuclide])
+    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]:
```suggestion
for nuclide in nuc_last:
    plot_color = random.choice(colors)
    time, num_dense[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
    print(time, num_dens[nuclide])
    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```

**Review by gonuke — CHANGES_REQUESTED** [review]
This is definitely getting closer.  I recommend more modularity overall in the OpenMC script.

File: VitaminJEnergyGroupStructure.xlsx
Comment (anu1217) [reply VitaminJEnergyGroupStructure.xlsx:1]: File has been deleted

File: myElelib
Comment (gonuke) [inline myElelib:1]: Is this the standard element library, or a customized one?  If the former, then we probably don't need it in the repo

File: fendl2bin.lib
Comment (gonuke) [inline fendl2bin.lib:1]: Don't need this file in the repo

File: SS_Activation_OpenMC.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+colors = list(mcolors.CSS4_COLORS.keys())
+num_dens= {}
+pair_list = {}
+
+for nuclide in nuc_last:
+    plot_color = random.choice(colors)
+    time, num_dens[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
+    print(time, num_dens[nuclide])
+    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: I'm noticing that your x-axis (time) on your plot goes up to just over 120,000 but you are depleting through 2.6e6 seconds... is something missing?

Comment (gonuke) [reply SS_Activation_OpenMC.py]: Related: I don't think the nuclides that remain at this time will experience much decay over 120,000 s.

File: SS_Activation_OpenMC.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+Tallies_CSV.to_csv('Tally_Values.csv', index=False)
+
+# Depletion results file
+results = openmc.deplete.Results(filename='depletion_results.h5')
+
+# Stable W nuclides present at beginning of operation (will not be plotted)
+stable_nuc = ['W180', 'W182', 'W183', 'W184', 'W186']  
+
+#Store list of nuclides from last timestep as a Materials object
+materials_last = results.export_to_materials(-1)
```
Comment (gonuke) [inline SS_Activation_OpenMC.py:131]: If you only plot the nuclides from the last time step, you might get relatively stable nuclides and might not see much decay.

File: SS_Activation_OpenMC.py (hunk 16 above)
Comment (gonuke) [reply SS_Activation_OpenMC.py]: Perhaps plot the y-axis on a log scale?

**Review by gonuke — CHANGES_REQUESTED** [review]
A few more comments while you work on this more.

File: myElelib
Comment (anu1217) [reply myElelib:1]: It's the library from the sample folder of ALARA. It's different from the standard one in that it has some other isotopes (e.g. Li-90) added in, but this doesn't matter for our W-only problem

File: SS_Activation_OpenMC.py [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+num_dens= {}
+pair_list = {}
+
+with open(r'Densities_CSV.csv', 'a') as density_file:
+
+    for nuclide in nuc_last:
+        plot_color = random.choice(colors)
+        time, num_dens[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
+        print(time, num_dens[nuclide])
+        density_file.write(''.join(map(str, nuclide)) + ',')
```
Comment (gonuke) [inline SS_Activation_OpenMC.py]: Since this is a single nuclide, this join approach is probably overkill
```suggestion
        density_file.write(f'{nuclide},')
```

**Review by gonuke — CHANGES_REQUESTED** [review]
Similar to the discussion of Eitan's PR yesterday - I wonder if we should focus on getting some of this merged and settled while you tweak the post-processing/comparisons.

Perhaps we finalize the ALARA and OpenMC input parts and separate it from the post-processing part.

File: SS_Activation_OpenMC.py (hunk 16 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py]: The graph is already on a log scale, but I was misinterpreting how OpenMC and Python track time.

Comment (anu1217) [reply SS_Activation_OpenMC.py]: I think the x-limits should be adjusted and the graph title changed to read "time after operation" rather than shutdown

File: SS_Activation_OpenMC.py (hunk 17 above)
Comment (anu1217) [reply SS_Activation_OpenMC.py:128]: The nuclides in the last timestep include the nuclides created at all earlier timesteps.

Comment (anu1217) [reply SS_Activation_OpenMC.py:128]: However, while I was playing around with the time_steps I found out something interesting. With the current operating time of 3e8 s, around 60 or so nuclides are at the last time step. If a small timestep is added at the beginning (e.g. 1000s) with the same neutron source rate, the number of nuclides at the end is around 230.

Comment (anu1217) [reply SS_Activation_OpenMC.py:128]: This seems to suggest that OpenMC isn't tracking the short-lived nuclides that are created near the beginning of operation, unless that short timestep is added in explicitly. Perhaps we should break up the operating time into smaller chunks to account for this

**Comment by anu1217** [comment]
> Similar to the discussion of Eitan's PR yesterday - I wonder if we should focus on getting some of this merged and settled while you tweak the post-processing/comparisons.
> 
> Perhaps we finalize the ALARA and OpenMC input parts and separate it from the post-processing part.

I was going to start a new pull request for the comparisons part, so I'm good with merging this if you are

**Review by gonuke — APPROVED** [review]

**Comment by gonuke** [comment]
I'm fine to merge this now, but we should discuss some code design issues for the next phase.
